File: 0108_Convert Sorted Array to Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sortedArrayToBST(self, nums):
        def helper(left, right):
            if left > right:
                return None  # Base case: no elements to process

            # Find the middle element to maintain balance
            mid = (left + right) // 2
            root = TreeNode(nums[mid])

            # Recursively build the left and right subtrees
            root.left = helper(left, mid - 1)
            root.right = helper(mid + 1, right)

            return root
        
        return helper(0, len(nums) - 1)

    # Optional: Helper function to print the tree for debugging
    def preOrder(self, node):
        if node is None:
            return
        print(node.val, end=" ")
        self.preOrder(node.left)
        self.preOrder(node.right)

# Example usage
if __name__ == "__main__":
    solution = Solution()

    nums1 = [-10, -3, 0, 5, 9]
    root1 = solution.sortedArrayToBST(nums1)
    print("Preorder traversal of the first BST:")
    solution.preOrder(root1)
    print()

    nums2 = [1, 3]
    root2 = solution.sortedArrayToBST(nums2)
    print("Preorder traversal of the second BST:")
    solution.preOrder(root2)
    print()

File: 0108_Convert Sorted Array to Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 public class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return helper(nums, 0, nums.length - 1);
    }

    private TreeNode helper(int[] nums, int left, int right) {
        if (left > right) {
            return null; // Base case: no elements to process
        }

        // Find the middle element to maintain balance
        int mid = left + (right - left) / 2;
        TreeNode root = new TreeNode(nums[mid]);

        // Recursively build the left and right subtrees
        root.left = helper(nums, left, mid - 1);
        root.right = helper(nums, mid + 1, right);

        return root;
    }

    // Optional: Helper function to print the tree for debugging
    public void preOrder(TreeNode node) {
        if (node == null) return;
        System.out.print(node.val + " ");
        preOrder(node.left);
        preOrder(node.right);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        int[] nums1 = {-10, -3, 0, 5, 9};
        TreeNode root1 = solution.sortedArrayToBST(nums1);
        System.out.println("Preorder traversal of the first BST:");
        solution.preOrder(root1);

        System.out.println();

        int[] nums2 = {1, 3};
        TreeNode root2 = solution.sortedArrayToBST(nums2);
        System.out.println("Preorder traversal of the second BST:");
        solution.preOrder(root2);
    }
}